File: krass_jass/trump.py
"""Rule-based trump selection.

`PLAN.md` §3.1 makes this the highest-value cheap component in the project: trump choice is
worth ~16 points of win rate over choosing at random, and a ranked rule-based selector comes
within ~0.7 points of a learned one. Search-based selection measured *worse*, largely
because it rarely learns to shove.

The scoring is deliberately transparent — weights live in `data/trump_weights.json` so they
can be tuned and reviewed without touching code.

**One thing here is easy to get backwards.** The contract multiplier scales the round's
points for *both* teams, so it multiplies your **edge**, not your score. Undenufe at ×4 with
a mediocre hand is a bad idea, not a good one worth four times as much. Contracts are
therefore compared on `(score - baseline) * multiplier`.
"""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path

from .cards import (
    NUM_RANKS,
    RANK_CHARS,
    SUIT_MASK,
    card_rank,
    count,
)
from .rules import SHOVE, Contract, RulesConfig

WEIGHTS_PATH = Path(__file__).parent / "data" / "trump_weights.json"


@lru_cache(maxsize=4)
def load_weights(path: str | None = None) -> dict:
    with open(path or WEIGHTS_PATH, encoding="utf-8") as fh:
        return json.load(fh)


def _rank_char(card: int) -> str:
    return RANK_CHARS[card_rank(card)]


def _suit_cards(hand: int, suit: int) -> list[int]:
    mask = hand & SUIT_MASK[suit]
    out = []
    while mask:
        low = mask & -mask
        out.append(low.bit_length() - 1)
        mask ^= low
    return out


def score_suit_as_trump(hand: int, suit: int, w: dict) -> float:
    """How good this hand is with `suit` as trump."""
    trumps = _suit_cards(hand, suit)
    score = sum(w["trump_rank_weights"][_rank_char(c)] for c in trumps)
    score += w["trump_length_bonus"][str(len(trumps))]

    for other in range(4):
        if other == suit:
            continue
        cards = _suit_cards(hand, other)
        score += sum(w["side_suit_weights"][_rank_char(c)] for c in cards)
        # Short side suits are ruffing chances, but only worth something with trumps
        # to ruff with — a void with two trumps is not the same as a void with six.
        if len(trumps) >= 3:
            if not cards:
                score += w["side_void_bonus"]["void"]
            elif len(cards) == 1:
                score += w["side_void_bonus"]["singleton"]
    return score


def _top_run_bonus(hand: int, w: dict, reverse: bool) -> float:
    """Consecutive top cards in a suit cash immediately in a no-trump contract.

    `reverse` walks from the 6 upward for Undenufe instead of from the ace down.
    """
    per_card = w["no_trump_top_run_bonus"]["per_card"]
    bonus = 0.0
    for suit in range(4):
        ranks = {card_rank(c) for c in _suit_cards(hand, suit)}
        order = range(NUM_RANKS - 1, -1, -1) if reverse else range(NUM_RANKS)
        for r in order:
            if r in ranks:
                bonus += per_card
            else:
                break
    return bonus


def score_obenabe(hand: int, w: dict) -> float:
    score = sum(
        w["obenabe_weights"][_rank_char(c)] for suit in range(4) for c in _suit_cards(hand, suit)
    )
    return score + _top_run_bonus(hand, w, reverse=False)


def score_undenufe(hand: int, w: dict) -> float:
    score = sum(
        w["undenufe_weights"][_rank_char(c)] for suit in range(4) for c in _suit_cards(hand, suit)
    )
    return score + _top_run_bonus(hand, w, reverse=True)


def score_all(hand: int, cfg: RulesConfig, weights: dict | None = None) -> dict[Contract, float]:
    """Edge-times-stakes score for every contract. Higher is better."""
    w = weights or load_weights()
    baseline = w["baseline"]["value"]

    raw = {Contract(s): score_suit_as_trump(hand, s, w) for s in range(4)}
    raw[Contract.OBENABE] = score_obenabe(hand, w)
    raw[Contract.UNDENUFE] = score_undenufe(hand, w)

    return {c: (score - baseline) * cfg.multiplier(c) for c, score in raw.items()}


def select_trump(
    hand: int,
    is_forehand: bool,
    cfg: RulesConfig,
    weights: dict | None = None,
) -> Contract | str:
    """Pick a contract, or `SHOVE`.

    Only forehand may shove, and only to a partner who must then choose — so this never
    returns `SHOVE` when `is_forehand` is false, or the bidding would not terminate.
    """
    w = weights or load_weights()
    scores = score_all(hand, cfg, w)
    best = max(scores, key=lambda c: (scores[c], -int(c)))

    if is_forehand and scores[best] < w["shove_threshold"]["value"]:
        return SHOVE
    return best


def describe(hand: int, cfg: RulesConfig) -> list[tuple[Contract, float]]:
    """Scores best-first, for tracing and for the `/how-it-works` documentation."""
    scores = score_all(hand, cfg)
    return sorted(scores.items(), key=lambda kv: -kv[1])

File: rust/src/trump.rs
//! Rule-based trump selection. Twin of `krass_jass/trump.py`.
//!
//! The weights are **compiled in from the same JSON** that the Python side reads
//! (`krass_jass/data/trump_weights.json`), because wasm has no filesystem. `include_str!`
//! keeps one source of truth: editing the JSON changes both implementations, and a
//! conformance test fails if they ever drift.
//!
//! The multiplier trap is the same one as in Python. It scales the round for *both* teams,
//! so it multiplies your **edge**, not your score — contracts are compared on
//! `(score - baseline) * multiplier`. Undenufe at ×4 with a mediocre hand is a bad idea,
//! not a good one worth four times as much.

use crate::cards::{card_rank, NUM_RANKS, NUM_SUITS, SUIT_MASK};
use crate::config::Rules;
use crate::tables::{NUM_CONTRACTS, OBENABE, UNDENUFE};

pub const SHOVE: usize = usize::MAX;

/// The weights file, parsed once at startup. Not a full JSON parser — it reads the exact
/// shape of the file it ships with, and a conformance test proves it agrees with Python.
const WEIGHTS_JSON: &str = include_str!("../../krass_jass/data/trump_weights.json");

pub struct Weights {
    pub trump_rank: [f64; NUM_RANKS],
    pub trump_length: [f64; NUM_RANKS + 1],
    pub side_suit: [f64; NUM_RANKS],
    pub obenabe: [f64; NUM_RANKS],
    pub undenufe: [f64; NUM_RANKS],
    pub void_bonus: f64,
    pub singleton_bonus: f64,
    pub top_run_per_card: f64,
    pub baseline: f64,
    pub shove_threshold: f64,
}

const RANK_KEYS: [&str; NUM_RANKS] = ["A", "K", "Q", "J", "T", "9", "8", "7", "6"];

fn find_object<'a>(text: &'a str, key: &str) -> &'a str {
    let at = text.find(&format!("\"{key}\"")).expect("weights key missing");
    let start = text[at..].find('{').expect("object expected") + at;
    let end = text[start..].find('}').expect("unterminated object") + start;
    &text[start..end]
}

fn number_after(block: &str, key: &str) -> Option<f64> {
    let at = block.find(&format!("\"{key}\""))?;
    let rest = &block[at + key.len() + 2..];
    let colon = rest.find(':')?;
    let tail = &rest[colon + 1..];
    let mut chars = tail.char_indices().skip_while(|(_, c)| c.is_whitespace());
    let (begin, _) = chars.next()?;
    let end = tail[begin..]
        .find(|c: char| c != '-' && c != '.' && !c.is_ascii_digit())
        .map(|i| begin + i)
        .unwrap_or(tail.len());
    tail[begin..end].trim().parse().ok()
}

fn rank_table(block: &str) -> [f64; NUM_RANKS] {
    let mut out = [0.0; NUM_RANKS];
    for (i, key) in RANK_KEYS.iter().enumerate() {
        out[i] = number_after(block, key).unwrap_or(0.0);
    }
    out
}

pub fn weights() -> &'static Weights {
    use std::sync::OnceLock;
    static CELL: OnceLock<Weights> = OnceLock::new();
    CELL.get_or_init(|| {
        let text = WEIGHTS_JSON;
        let length_block = find_object(text, "trump_length_bonus");
        let mut trump_length = [0.0; NUM_RANKS + 1];
        for (i, slot) in trump_length.iter_mut().enumerate() {
            *slot = number_after(length_block, &i.to_string()).unwrap_or(0.0);
        }
        let voids = find_object(text, "side_void_bonus");
        Weights {
            trump_rank: rank_table(find_object(text, "trump_rank_weights")),
            trump_length,
            side_suit: rank_table(find_object(text, "side_suit_weights")),
            obenabe: rank_table(find_object(text, "obenabe_weights")),
            undenufe: rank_table(find_object(text, "undenufe_weights")),
            void_bonus: number_after(voids, "void").unwrap_or(0.0),
            singleton_bonus: number_after(voids, "singleton").unwrap_or(0.0),
            top_run_per_card: number_after(find_object(text, "no_trump_top_run_bonus"), "per_card")
                .unwrap_or(0.0),
            baseline: number_after(find_object(text, "baseline"), "value").unwrap_or(0.0),
            shove_threshold: number_after(find_object(text, "shove_threshold"), "value")
                .unwrap_or(0.0),
        }
    })
}

fn suit_cards(hand: u64, suit: usize) -> Vec<usize> {
    let mut out = Vec::new();
    let mut mask = hand & SUIT_MASK[suit];
    while mask != 0 {
        let low = mask & mask.wrapping_neg();
        out.push(low.trailing_zeros() as usize);
        mask ^= low;
    }
    out
}

pub fn score_suit_as_trump(hand: u64, suit: usize, w: &Weights) -> f64 {
    let trumps = suit_cards(hand, suit);
    let mut score: f64 = trumps.iter().map(|&c| w.trump_rank[card_rank(c)]).sum();
    score += w.trump_length[trumps.len()];

    for other in 0..NUM_SUITS {
        if other == suit {
            continue;
        }
        let cards = suit_cards(hand, other);
        score += cards.iter().map(|&c| w.side_suit[card_rank(c)]).sum::<f64>();
        // Short side suits are ruffing chances, but only with trumps to ruff with.
        if trumps.len() >= 3 {
            if cards.is_empty() {
                score += w.void_bonus;
            } else if cards.len() == 1 {
                score += w.singleton_bonus;
            }
        }
    }
    score
}

fn top_run_bonus(hand: u64, w: &Weights, reverse: bool) -> f64 {
    let mut bonus = 0.0;
    for suit in 0..NUM_SUITS {
        let held: Vec<usize> = suit_cards(hand, suit).into_iter().map(card_rank).collect();
        let order: Vec<usize> = if reverse {
            (0..NUM_RANKS).rev().collect()
        } else {
            (0..NUM_RANKS).collect()
        };
        for r in order {
            if held.contains(&r) {
                bonus += w.top_run_per_card;
            } else {
                break;
            }
        }
    }
    bonus
}

pub fn score_obenabe(hand: u64, w: &Weights) -> f64 {
    let mut score = 0.0;
    for suit in 0..NUM_SUITS {
        score += suit_cards(hand, suit).iter().map(|&c| w.obenabe[card_rank(c)]).sum::<f64>();
    }
    score + top_run_bonus(hand, w, false)
}

pub fn score_undenufe(hand: u64, w: &Weights) -> f64 {
    let mut score = 0.0;
    for suit in 0..NUM_SUITS {
        score += suit_cards(hand, suit).iter().map(|&c| w.undenufe[card_rank(c)]).sum::<f64>();
    }
    score + top_run_bonus(hand, w, true)
}

/// Edge-times-stakes score for every contract. Higher is better.
pub fn score_all(hand: u64, rules: &Rules) -> [f64; NUM_CONTRACTS] {
    let w = weights();
    let mut out = [0.0; NUM_CONTRACTS];
    for suit in 0..NUM_SUITS {
        out[suit] = score_suit_as_trump(hand, suit, w);
    }
    out[OBENABE] = score_obenabe(hand, w);
    out[UNDENUFE] = score_undenufe(hand, w);
    for (contract, slot) in out.iter_mut().enumerate() {
        *slot = (*slot - w.baseline) * rules.multiplier(contract) as f64;
    }
    out
}

/// Pick a contract, or `SHOVE`. Never shoves off forehand, or the bidding would not
/// terminate.
pub fn select_trump(hand: u64, is_forehand: bool, rules: &Rules) -> usize {
    let scores = score_all(hand, rules);
    let mut best = 0usize;
    for contract in 1..NUM_CONTRACTS {
        // Ties go to the lower contract index, matching Python's `(score, -index)` key.
        if scores[contract] > scores[best] {
            best = contract;
        }
    }
    if is_forehand && scores[best] < weights().shove_threshold {
        return SHOVE;
    }
    best
}
